common/utils/query_utils.ts

```diff
@@ -23,14 +23,84 @@ const escapeQuotes = (literal: string) => {
   return literal.replaceAll("'", "''");
 };
 
-export const getIndexPatternFromRawQuery = (query: string): string => {
+const PROMQL_INDEX_REGEX = /(?<connection>[^|\s]+)\.query_range\('(?<promql>.+?)'/i;
+const PROMQL_AGG_ATTR_REGEX = /(?<agg>\S+) *?(by\((?<attrs>.*?)\))? *?\((?<metric>\S+)\)/i;
+const PROMQL_METRIC_ONLY_REGEX = /^[(^\s]*?(?<metric>[_.a-z0-9]+)\)?/i;
+
+const promQlMetricOnly = (promQL: string) => promQL.match(PROMQL_METRIC_ONLY_REGEX)?.groups;
+const promQLAggAttr = (promQL: string) => promQL.match(PROMQL_AGG_ATTR_REGEX)?.groups;
+
+interface PromQLAST {
+  connection: string;
+  metric: string;
+  agg?: string;
+  attrs?: string;
+}
+
+const promQLMatch = (query: string): PromQLAST | undefined => {
+  const indexMatch = query.match(PROMQL_INDEX_REGEX);
+  if (!indexMatch) return;
+  const { connection, promql } = indexMatch.groups;
+  if (!connection || !promql) return;
+
+  const promAst = promQLAggAttr(promql) || promQlMetricOnly(promql);
+  if (!promAst) return;
+
+  return { ...promAst, connection };
+};
+
+export const getPromQLIndex = (query: string): string | undefined => {
+  const match = promQLMatch(query);
+  if (!match) return;
+
+  const { connection, metric } = match!;
+  if (connection && metric) return connection + '.' + metric;
+};
+
+export const isPromQLQuery = (query: string): boolean => !!promQLMatch(query);
+
+export const buildPromQLFromMetricQuery = (query: MetricType['query']): string => {
+  const { catalog, aggregation, attributesGroupBy } = query;
+  const [connection, metric] = catalog.split('.');
+
+  const attrs = attributesGroupBy.length > 0 ? `by (${attributesGroupBy.join(',')})` : '';
+
+  console.log('buildPromQLFromMetricQuery', { catalog, aggregation, attrs, metric });
+  return `${aggregation} ${attrs} (${metric})`;
+};
+
+const getPPLIndex = (query: string): string => {
   const matches = query.match(PPL_INDEX_REGEX);
   if (matches) {
     return matches[2];
   }
   return '';
 };
 
+export const getIndexPatternFromRawQuery = (query: string): string => {
+  return getPromQLIndex(query) || getPPLIndex(query);
+};
+
+const preProcessPromQLQuery = ({
+  promQLAst,
+  startTime,
+  endTime,
+  timeField,
+}: {
+  promQLAst: PromQLAST;
+  startTime: string;
+  endTime: string;
+  timeField?: string;
+}): string => {
+  const { connection, metric } = promQLAst;
+
+  // convert to epoch-sec
+  const start = datemath.parse(startTime)?.utc()?.unix();
+  const end = datemath.parse(endTime, { roundUp: true })?.utc()?.unix();
+
+  return `source=${connection}.query_range('${metric}', ${start}, ${end}, '1d')`;
```

Comment:
would it be possible for `datemath.parse()` to return undefined? If so could the code check for it to prevent having `undefined` in the returned string rather than the `?.utc()?.unix()` chaining?

Reply:
Corrected.  Calling common convertDateTime()